### PR TITLE
HDDS-9131. [hsync] A freon tool to focus on hsync/hflush performance

### PR DESCRIPTION
**Pull request by jojochuang**
## What changes were proposed in this pull request?

Add a -sync option for ozone freon dfsg which does hsync/hflush after write.

Example:

$ ozone freon dfsg --path ofs://ozone1/vol1/bucket1/a1 -s 1024 --buffer 1024 -sync hflush -t 8 -n 10000
starts 8 threads to write 10000 files, each file is hflush'ed when reaching the end of buffer.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9131

## How was this patch tested?

Unit tests, manual tests

### CODE REVIEW

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopFsGenerator.java [hunk 1]
```diff
@@ -63,6 +63,12 @@ public class HadoopFsGenerator extends BaseFreonGenerator
       defaultValue = "4096")
   private int copyBufferSize;
 
+  @Option(names = {"-sync"},
+      description = "Type of operation to execute after a write. Supported " +
+      "options include none (default), hflush and hsync",
+      defaultValue = "none")
+  private String flushOrSync;
```
Comment (duongkame) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopFsGenerator.java:70]: Think Option works well with enum types, e.g. https://github.com/apache/ozone/blob/master/hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/ListSubcommand.java#L72-L74

Exposing `SyncOptions` as a top-level type and using it directly here looks simpler.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopFsGenerator.java [hunk 2]
```diff
@@ -63,6 +63,12 @@ public class HadoopFsGenerator extends BaseFreonGenerator
       defaultValue = "4096")
   private int copyBufferSize;
 
+  @Option(names = {"-sync"},
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopFsGenerator.java:66]: Please avoid single-dash long options.

```suggestion
  @Option(names = {"--sync"},
```

**Comment by duongkame** [comment]
CI's failing, UT:
```
 Tests run: 6, Failures: 0, Errors: 6, Skipped: 0, Time elapsed: 0.807 s <<< FAILURE! - in org.apache.hadoop.ozone.freon.TestContentGenerator
Error:  org.apache.hadoop.ozone.freon.TestContentGenerator.writeWrite  Time elapsed: 0.015 s  <<< ERROR!
java.lang.NullPointerException
	at org.apache.hadoop.ozone.freon.ContentGenerator.doFlushOrSync(ContentGenerator.java:106)
	at org.apache.hadoop.ozone.freon.ContentGenerator.write(ContentGenerator.java:99)
	at org.apache.hadoop.ozone.freon.TestContentGenerator.writeWrite(TestContentGenerator.java:42)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

```

**Comment by jojochuang** [comment]
Fixing that now

**Comment by jojochuang** [comment]
The failed test was only merged in master a few hours ago and doesn't look related.

**Comment by jojochuang** [comment]
Thanks @adoroszlai . I am curious how did you merge commits and update my PR :) Can any committers update someone else's PR?

**Comment by adoroszlai** [comment]
> I am curious how did you merge commits and update my PR :) Can any committers update someone else's PR?

When creating the PR, there is a checkbox `Allow edits and access to secrets by maintainers`, enabled by default.  If it is left enabled, committers can update the PR.

**Review by duongkame — APPROVED** [review]
LGTM. +1

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopFsGenerator.java [hunk 3]
```diff
@@ -63,6 +63,12 @@ public class HadoopFsGenerator extends BaseFreonGenerator
       defaultValue = "4096")
   private int copyBufferSize;
 
+  @Option(names = {"--sync"},
+      description = "Type of operation to execute after a write. Supported " +
+      "options include NONE (default), HFLUSH and HSYNC",
+      defaultValue = "none")
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HadoopFsGenerator.java:69]:
```suggestion
      defaultValue = "NONE")
```

`none` gives error if the option is omitted:

```
$ ozone freon dfsg --path ofs://om/vol1/buck1/dir1 -s 1024 -t1 -n1
...
Invalid value for option '--sync': expected one of [NONE, HFLUSH, HSYNC] (case-sensitive) but was 'none'
```

**Review by adoroszlai — CHANGES_REQUESTED** [review]
Thanks @jojochuang for updating the patch.

**Review by adoroszlai — APPROVED** [review]

**Comment by adoroszlai** [comment]
Thanks @jojochuang for the patch, @duongkame for the review.

**Pull request merged**